### PR TITLE
Add calder2 module

### LINKED ISSUE #2293: new module: calder2
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by lucananni93**
## PR checklist

Closes #2293 

- [X] This comment contains a description of changes (with reason).
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [X] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [X] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [X] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/calder2/main.nf [hunk 1]
```diff
@@ -0,0 +1,43 @@
+def VERSION = '0.3'
```
Comment (jfy133) [inline modules/nf-core/calder2/main.nf:1]: This needs to be in the script block: https://github.com/nf-core/modules/blob/f919110f52861a5c0c2ab8229c17560f41405721/modules/nf-core/homer/annotatepeaks/main.nf#L26

File: modules/nf-core/calder2/main.nf [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def args = task.ext.args ?: ''
+    def suffix = resolution ? "::$resolution" : ""
+    def cpus = task.cpus ?: 1
+    """
+    # getting binsize
+    binsize="\$(cooler info --field bin-size $cool$suffix)"
+
+    echo "Detected binsize: \${binsize}"
+
+    # calling CALDER
```
Comment (jfy133) [inline modules/nf-core/calder2/main.nf:35]:
```suggestion
```

File: modules/nf-core/calder2/main.nf [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def suffix = resolution ? "::$resolution" : ""
+    def cpus = task.cpus ?: 1
+    """
+    # getting binsize
```
Comment (jfy133) [inline modules/nf-core/calder2/main.nf:30]:
```suggestion
    # getting binsize as mandatory input for calder
```

File: modules/nf-core/calder2/main.nf [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    binsize="\$(cooler info --field bin-size $cool$suffix)"
+
+    echo "Detected binsize: \${binsize}"
+
+    # calling CALDER
+    calder --input $cool$suffix --outpath ${meta.id} --nproc $cpus --type cool --bin_size "\${binsize}" $args
```
Comment (jfy133) [inline modules/nf-core/calder2/main.nf:36]: - Please split each parameter across multiple lines (see other modules as example)
- Please use `prefix` not `meta.id` for outpath, you will see examples for this in other modules [^1]


[^1]: https://nf-co.re/docs/contributing/modules#naming-conventions point 6

File: modules/nf-core/calder2/meta.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+name: "calder2"
+description: Hierarchical Hi-C compartment computation
+keywords:
+  - calder2
```
Comment (jfy133) [inline modules/nf-core/calder2/meta.yml:4]: More keywords please! make the module searchable, e.g. `hi-c` etc. you can see more examples/suggestions in the green tags here: https://nf-co.re/modules

File: modules/nf-core/calder2/meta.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+output:
+  - meta:
+      type: map
+      description: Groovy Map containing sample information. E.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - output:
+      type: folder
```
Comment (jfy133) [inline modules/nf-core/calder2/meta.yml:35]:
```suggestion
      type: directory
```

File: modules/nf-core/calder2/meta.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+      description: Calder output
```
Comment (jfy133) [inline modules/nf-core/calder2/meta.yml:36]: Containing....?

File: tests/modules/nf-core/calder2/main.nf [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+def TEST_COOL_PATH = "https://raw.githubusercontent.com/CSOgroup/CALDER2/main/tests/testthat/data/test.cool"
+
+
+workflow test_calder2 {
+    
+    input = [ [ id:'test' ], //meta map
+              file(TEST_COOL_PATH, checkIfExists: true) ]
+
+    CALDER2 ( input, [:] )
+}
```
Comment (jfy133) [inline tests/modules/nf-core/calder2/main.nf:17]: Can you add a test also using the `resolution` input channel?

**Review by jfy133 — CHANGES_REQUESTED** [review]

**Comment by lucananni93** [comment]
@jfy133 Is it possible to merge now?

File: modules/nf-core/calder2/meta.yml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  - meta:
+      type: map
+      description: Groovy Map containing sample information. E.g. [ id:'test', single_end:false ]
+  - input:
+      type: file
+      description: Path to COOL file
+      pattern: "*.{cool.mcool}"
+  - resolution:
+      type: value
+      description: Resolution
```
Comment (jfy133) [inline modules/nf-core/calder2/meta.yml:29]: Please expand on 'resolution' here

**Review by jfy133 — APPROVED** [review]
One more docs thing,and you're good to go.

I also see there is a `--keep_intermediate` flag, you may wish to make an optional output channel to pick up these files if someone supplies that flag via modules.conf or similar. However this is not a dealbreak for me.

But otherwise LGTM

```
Usage: /home/jfellows/.conda/envs/calder2/bin/calder [options]


Options:
	-i INPUT, --input=INPUT
		Input Hi-C contacts

	-t TYPE, --type=TYPE
		The type of input: hic or cool [default hic]

	-b BIN_SIZE, --bin_size=BIN_SIZE
		Bin size to use for the analysis [default 50000]

	-g GENOME, --genome=GENOME
		Genome assembly to use [default hg19]

	-f FEATURE_TRACK, --feature_track=FEATURE_TRACK
		Genomic feature track to be used to determine A/B compartment direction
	              when genome == 'others'. The track should presumably have higher values
	              in A than in B compartmnets. [default NA]

	-p NPROC, --nproc=NPROC
		Number of cores to use [default 1]

	-o OUTPATH, --outpath=OUTPATH
		Path to the output folder

	-k, --keep_intermediate
		Keep intermediate data after done [default FALSE]

	-h, --help
		Show this help message and exit
```

**Comment by lucananni93** [comment]
I've added the optional folder for intermediate data to the module and improved the description of the "Resolution" output.
All checks are successful. If you agree, @jfy133, I'll proceed with the merge.

**Review by jfy133 — APPROVED** [review]
Yes you can merge

**Pull request merged**